ctapipe/reco/shower_processor.py

```diff
@@ -108,23 +99,18 @@ def _reconstruct_shower(
             list of tel_ids used if stereo, or None if Mono
         """
 
-        # Read only valid HillasContainers (minimum condition to continue)
+        # Select only images which pass the shower quality criteria
         hillas_dict = {
             tel_id: dl1.parameters.hillas
             for tel_id, dl1 in event.dl1.tel.items()
-            if np.isfinite(event.dl1.tel[tel_id].parameters.hillas.intensity)
+            if all(self.check_shower(dl1.parameters))
         }
-
-        # On top of this check if the shower should be considered based
-        # on the user's configuration
-        shower_criteria = [self.check_shower(hillas_dict[tel_id]) for tel_id in hillas_dict]
         self.log.debug(
-            "shower_criteria: %s",
-            list(zip(self.check_shower.criteria_names[1:], shower_criteria)),
+            f"shower_criteria:\n {self.check_shower.to_table()}"
```

Comment:
Does this log the shower info for all the events all the time? This might be quite expensive since `to_table` is also called when the level is higher than debug.

Reply:
Yes, it should be for all events.
I chose this solution because the fancier hillas_dict uses `self.check_shower` on a telescope-wise quantity, whereas the old syntax on the event, so I didn't want to define `shower_criteria` only for the debug print

Reply:
technically it should be the `quality_by_tel` quantity that Karl proposed, but it seemed overly complicated just for the print

Reply:
Since this is an f-string, it's evaluated for every event, which is bad for performance.  If you change it to:
```python
self.log.debug("shower_criteria:\n %s", self.check_shower.to_table())
```
The call to `to_table()` will only be made if the log-level is debug, I think.  

However, if this is useful information, we should consider just adding it to the Container so it's written out for each event, rather than a debug statement.  If you add a field like "quality_flags", it would get written as a boolean vector column.

Reply:
Though this is telescope-wise information, not really shower-wise no?

I thought we decided not to add telescope-wise data to DL2 (I guess the container address you are referring to is `event.dl2.shower["HillasReconstructor"]` for example)

When we print `self.check_shower.to_table()` we obtain something like,

```
   criteria    counts cumulative_counts
-------------- ------ -----------------
         TOTAL      8                 8
      > 50 phe      8                 8
Positive width      8                 8
    > 3 pixels      8                 8
```
but what we really use to filter the hillas dictionaries is 

`self.check_shower(dl1.parameters)`

which is something like `[ True  True  True]` for each `tel_id`

so what do we want to save?

I'd say that a dict of dicts like `{tel_id: {cut_name: bool}}` would be the most informative choice, but the container we are talking about here is DL2...

Reply:
For the moment I changed the debug logging to a normal `%` format string

Reply:
You actually want to save two things: for each event you want to save the results of check_shower into a Field (it will become a column of boolean arrays when written), and also at the end you want to write the to_table() information to the service group as we did in the DL1 part (that contains overall statistics, but not event-wise information). 

So add a Field to the container like `quality_flags`, and place the output of `check_shower(params)`  there.

Reply:
But what you call "boolean arrays" tells you if an image passed **all** the checks, not which specific cuts

Reply:
It does, you store each flag for each event (a boolean array per event, resulting in a 2d (n_events, n_checks) column)

Reply:
but here by event you mean the shower or the image?
because we want to write this to ReconstructedShowerContainer

Reply:
No, this needs to be written per telescope.

Reply:
I mean, when you do `self.check_shower(dl1.parameters)`

you get for example

`[True False True]`

which means there are 3 images and only 1 was not used, because **one** of the cuts was not passed , but you cannot know which one from this information

Reply:
ah no sorry

Reply:
this is for 1 shower, with 3 images and each boolean is 1 cut from the quality query - right?

Reply:
then by using a Map we should link the tel_ids with each cut-related boolean, am I correct?

Reply:
You may want to wait until we merge #1673 since that updates the internal data model to have a place to put per-telescope DL2 data, which is missing in this version.  Then you can add this boolean there - I just realized that may be what is causing confusion.

Reply:
yes you're right. I'll try to review again that so we can proceed!